pkgs/test_core/lib/src/runner/vm/test_compiler.dart

```diff
@@ -0,0 +1,273 @@
+  final _compilationQueue = <_CompilationRequest>[];
+  final _stdoutHandler = _StdoutHandler();
+
+  File? _outputDill;
+  Process? _compiler;
+  PackageConfig? _packageConfig;
+
+  Future<String?> compile(Uri mainDart, Metadata metadata) {
+    final completer = Completer<String>();
+    if (_compilerController.isClosed) {
+      return Future.value(null);
+    }
+    _compilerController.add(_CompilationRequest(mainDart, completer, metadata));
+    return completer.future;
+  }
+
+  Future<void> _shutdown() async {
+    if (_compiler != null) {
+      _compiler!.kill();
+      _compiler = null;
+    }
+  }
+
+  Future<void> dispose() async {
+    await _compilerController.close();
+    await _shutdown();
+  }
+
+  Future<String> _languageVersionComment(Uri testUri) async {
+    var localPackageConfig = _packageConfig ??= await loadPackageConfig(File(
+        path.join(
+            Directory.current.path, '.dart_tool', 'package_config.json')));
+    var package = localPackageConfig.packageOf(testUri) as Package?;
+    if (package == null) {
+      return '';
+    }
+    return '// @dart=${package.languageVersion.major}.${package.languageVersion.minor}';
+  }
+
+  Future<String> _generateEntrypoint(
+      Uri testUri, Metadata suiteMetadata) async {
+    return '''
+        ${suiteMetadata.languageVersionComment ?? await _languageVersionComment(testUri)}
+    import "dart:isolate";
+
+    import "package:test_core/src/bootstrap/vm.dart";
+
+    import "$testUri" as test;
+
+    void main(_, SendPort sendPort) {
+      internalBootstrapVmTest(() => test.main, sendPort);
+    }
+  ''';
+  }
+
+  // Handle a compilation request.
+  Future<void> _onCompilationRequest(_CompilationRequest request) async {
+    final isEmpty = _compilationQueue.isEmpty;
+    _compilationQueue.add(request);
+    if (!isEmpty) {
+      return;
+    }
+    while (_compilationQueue.isNotEmpty) {
+      final request = _compilationQueue.first;
+      var firstCompile = false;
+      _CompilerOutput? compilerOutput;
+      final contents =
+          await _generateEntrypoint(request.mainUri, request.metadata);
+      final tempFile = File(path.join(_outputDillDirectory.path, 'test.dart'))
+        ..writeAsStringSync(contents);
+
+      if (_compiler == null) {
+        compilerOutput = await _createCompiler(tempFile.uri);
+        firstCompile = true;
+      } else {
+        compilerOutput = await _recompile(
+          tempFile.uri,
+        );
+      }
+      final outputPath = compilerOutput?.outputFilename;
+      if (outputPath == null || compilerOutput!.errorCount > 0) {
+        request.result.complete(null);
+        await _shutdown();
+      } else {
+        final outputFile = File(outputPath);
+        final kernelReadyToRun = await outputFile.copy('${tempFile.path}.dill');
+        final testCache = File(_dillCachePath);
+        if (firstCompile ||
+            !testCache.existsSync() ||
+            (testCache.lengthSync() < outputFile.lengthSync())) {
```

Comment:
Worth commenting on this condition - is the assumption that larger file == more relevant for future runs basically?